bundler/lib/bundler/definition.rb

```diff
@@ -310,34 +310,24 @@ def groups
       dependencies.map(&:groups).flatten.uniq
     end
 
-    def lock(file, preserve_unknown_sections = false)
-      return if Definition.no_lock
-
-      contents = to_lock
-
-      # Convert to \r\n if the existing lock has them
-      # i.e., Windows with `git config core.autocrlf=true`
-      contents.gsub!(/\n/, "\r\n") if @lockfile_contents.match?("\r\n")
-
-      if @locked_bundler_version
-        locked_major = @locked_bundler_version.segments.first
-        current_major = bundler_version_to_lock.segments.first
-
-        updating_major = locked_major < current_major
-      end
-
-      preserve_unknown_sections ||= !updating_major && (Bundler.frozen_bundle? || !(unlocking? || @unlocking_bundler))
+    def lock(file_or_preserve_unknown_sections = false, preserve_unknown_sections_or_unused = false)
+      if [true, false, nil].include?(file_or_preserve_unknown_sections)
+        target_lockfile = lockfile || Bundler.default_lockfile
+        preserve_unknown_sections = file_or_preserve_unknown_sections
+      else
+        target_lockfile = file_or_preserve_unknown_sections
+        preserve_unknown_sections = preserve_unknown_sections_or_unused
 
-      return if file && File.exist?(file) && lockfiles_equal?(@lockfile_contents, contents, preserve_unknown_sections)
+        suggestion = if target_lockfile == lockfile
+          "To fix this warning, remove it from the `Definition#lock` call."
+        else
+          "Instead, instantiate a new definition passing `#{target_lockfile}`, and call `lock` without a file argument on that definition"
```

Comment:
Maybe a wild idea, but in this case, we could simplify Definition by literally re-creating the Definition on this branch and calling lock, just as we're advising. Then definition would only ever write to the lockfile given at the start. It would be a performance hit but hopefully no one fixes the warning and it would maybe allow us to commit to a "final" version for the rest of the Definition API! Too crazy?

Reply:
Not crazy at all! Neat idea on the contrary. Will update 👍.

Reply:
Nice. When I reread my comment I probably meant "hopefully no one ignores the warning". I'm guessing the performance hit won't matter because it will get fixed, then we can eventually remove this code branch.

Reply:
Yes, I was confused too by that part, but I assumed you meant exactly the opposite :) I agree performance is not a concern in deprecated code paths.